sams-scala-tutorial/serialization-exceptions-and-memory-leaks-no-ws.md

```diff
@@ -13,14 +13,147 @@ Talk:
 
 ## What causes serialization problems
 
-Two things:
+Unserializable data trnasferred across nodes, there are 2 kinds of problems:
 
 1. Spark jobs that *may* require transfer of *object* data between JVMs
     - Trying to broadcast non-serializable data
     - Trying to collect non-serializable data
     - Trying to *shuffle* non-serializable data
 2. Spark jobs that try to transfer non-serializable "*meta*" data between JVMs via (intentional or unintentional) **transative closure capture**.
 
+## Examples for problem 2 and solutions
+```
+object TestSerializable {
+  def main(args: Array[String]): Unit = {
+    val sparkConf = new SparkConf
+    val sc = new SparkContext(sparkConf)
+
+    val data = List(1, 2, 3, 4, 5)
+    val rdd = sc.makeRDD(data)
+
+    /* What is the unserializable task?
+     * When some methods applied in the RDD closure, everything necessary to execute the closure will be packaged up, serialized and sent to executors.
+     * In this process, all functions would be passed into a process called ClosureCleaner. 
+     * If the functions in the closure cannot be cut off from its context, there will be an exception called NotSerializableException.
+     * */
+
+    val unSerializableClassHandler = new NonSerializableClass(1)
+    rdd.map { x => unSerializableClassHandler.func_1 }.count()
+
```

Comment:
Hello Sam, you are right. directly calling object's method in `map()` will solve this problem and I wrote this solution in the following in solution 3. The reason that I create this example is to illustrate one reason unserialization issue may happen, that is when we want to use a method in object created in driver, every thing in the object need to be seriailized. In this example, object is not serializable so the TSNE happens.